pip-0002.md

```diff
@@ -0,0 +1,58 @@
+# PIP-2: Perspective View Web Component Interface spec.
+
+AD4M [Perspectives](https://github.com/perspect3vism/ad4m/blob/main/src/perspectives/Perspective.ts) can be rendered and interacted with in many different ways.
+
+The thinking behind AD4M is that there can and should be many different UIs to visualized all or parts of the same Perspective. These different UIs can take the form of applications that connect to the AD4M-executor and select *their* Perspective(s) to visualize.
+
+But already for both first AD4M applications, Junto's [Flux](https://github.com/juntofoundation/flux) and [Perspect3ve](https://github.com/perspect3vism/perspect3ve), we are dealing with multiple Prespectives (all Perspectives in the case of Perspect3ve) that should be displayed and interacted with in different ways: 
+ * as graph
+ * as chat thread
+ * as plain list of links for debugging
+ * ...
+
+There is nothing conceptually stopping us from regarding these different ways of displaying of and interacting with Perspectives as re-usable componentes, that are largely independent from the actual app in which they are used (Flux / Perspect3ve / ...) and thus could be build as web components and shared between apps.
+
+For that to pragmatically work out, we have to agree on an interface for these web components - that is agree on their **properties**, **events** and **methods**.
+
+## Properties
+
+- `perspective-uuid`: *string*
+- `ad4m-client`: *Ad4mClient*
+
+The Perspective View component needs to be initialized with the Perspective to render and with access to it.
+
+The combination of the Perspective's ID and an initialized [`Ad4mClient`](https://github.com/perspect3vism/ad4m/blob/main/src/Ad4mClient.ts) object is all that's needed since the component can retrieve a PerspectiveProxy object via the client with the given ID and then do all queries and mutation through that proxy.
+
+There was some discussion around passing in an `Ad4mClient` or having the component create one. We decided for passing it in and thus leaving the concern about matching versions between `Ad4mClient` and the AD4M-executor to the app that uses the Perspective View component. Otherwise updates to the AD4M-executor interface would potentially break Perspective View components and force apps to either wait for the component developer to update their code, or stick to an old version to keep the used Perspective View(s).
```

Comment:
I just thought some more about this and I think I prefer passing in just a `PerspectiveProxy`.

Benefits:
* we don't have to pass in nor create an `Ad4mClient` inside the view, the proxy object already includes/wraps an Ad4mClient, but only offers methods for working with the specified perspective, which leads to the 2nd point:
* security - or at least the possibility of restricting access for a perspective view.

For instance: we could have a read-only PerspectiveProxy, or one that even filters certain links/information.

We might want that in some cases since downloading, installing and running a perspective view means running code from a foreign developer that would have access to all my perspectives. We will need a similar install and provenance checking mechanic like with the Languages, but since Perspective Views don't *have to* have access to everything since they are running in the UI in the first place and only dealing with one Perspective, it makes sense to *try out* a view you've found somewhere but only use it on a non-critical perspective...

Thoughts?

Reply:
There could be more optional properties for the other parts of Ad4mClient, which makes it an explicit decision to grant more access to the Perspective View, if needed.

Reply:
@leifriksheim, I will change it over to PerspectiveProxy now and would like to merge this soon. I also want to create a next PIP for virtual/custom icons for expressions and I have tested some ideas there as well. Restricting access to exactly the pieces needed for that component makes more and more sense to me. Any comments?

Reply:
What about if I want to render expressions that I find in an Perspective's link? Seems like a pretty standard use case that would not be covered by only having the PerspectiveProxy interactions available.

Perhaps this is where having sessions built into the ad4m-executor comes into play? Whereby the ad4mClient can init multiple sessions each with different permissions scopes to be delegated to given perspective views?

Reply:
Had the same need with the virtual icons and I've added the ExpressionClient as another prop there. Would do the same here. Will update this PIP today.

Reply:
This is the props I used in the first virtual icon:
```javascript
    export let base: string
    export let perspective: PerspectiveProxy
    export let expressionClient: ExpressionClient
    export let me: Agent
    export let includeBase: boolean
```

Reply:
Okay that makes sense. Lets also think about another PIP for some session mechanism and the permissions for that session.